mlx/backend/metal/fft.cpp

```diff
@@ -1,12 +1,113 @@
 // Copyright © 2023 Apple Inc.
 
+#include "mlx/backend/metal/copy.h"
+#include "mlx/backend/metal/utils.h"
+#include "mlx/mlx.h"
 #include "mlx/primitives.h"
 
 namespace mlx::core {
 
 void FFT::eval_gpu(const std::vector<array>& inputs, array& out) {
+  auto& s = out.primitive().stream();
+  auto& d = metal::device(s.device);
+
   auto& in = inputs[0];
-  throw std::runtime_error("[FFT] NYI for Metal backend.");
+
+  if (axes_.size() == 0 || axes_.size() > 1 || inverse_ ||
+      in.dtype() != complex64 || out.dtype() != complex64) {
+    // Could also fallback to CPU implementation here.
+    throw std::runtime_error(
+        "GPU FFT is only implemented for 1D, forward, complex FFTs.");
+  }
+
+  size_t n = in.shape(axes_[0]);
+
+  if (!is_power_of_2(n) || n > 2048 || n < 4) {
+    throw std::runtime_error(
+        "GPU FFT is only implemented for the powers of 2 from 4 -> 2048");
+  }
+
+  // Make sure that the FFT dimension is contiguous
+  std::vector<array> copies;
+  auto check_input = [this, &copies, &s](const array& x) {
+    bool no_copy = x.strides()[axes_[0]] == 1;
+    if (no_copy) {
+      return x;
+    } else {
+      // Set stride at FFT dim to 1 and fill in others
+      array x_copy(x.shape(), x.dtype(), nullptr, {});
+      std::vector<size_t> strides;
+      size_t cur_stride = x.shape(axes_[0]);
+      for (int axis = 0; axis < x.ndim(); axis++) {
+        if (axis == axes_[0]) {
+          strides.push_back(1);
+        } else {
+          strides.push_back(cur_stride);
+          cur_stride *= x.shape(axis);
+        }
+      }
+
+      // set the contiguity flags
+      auto flags = x.flags();
+      size_t f_stride = 1;
+      size_t b_stride = 1;
+      flags.col_contiguous = true;
+      flags.row_contiguous = true;
+      for (int i = 0, ri = x.ndim() - 1; i < x.ndim(); ++i, --ri) {
+        flags.col_contiguous &= (strides[i] == f_stride || x.shape(i) == 1);
+        f_stride *= x.shape(i);
+        flags.row_contiguous &= (strides[ri] == b_stride || x.shape(ri) == 1);
+        b_stride *= x.shape(ri);
+      }
+      // FFT dim has stride 1 so there are no gaps
+      flags.contiguous = true;
+
+      x_copy.set_data(
+          allocator::malloc_or_wait(x.nbytes()), x.data_size(), strides, flags);
+      copy_gpu_inplace(x, x_copy, CopyType::GeneralGeneral, s);
+      copies.push_back(x_copy);
+      return x_copy;
+    }
+  };
+  const array& in_contiguous = check_input(inputs[0]);
```

Comment:
This array, is it actually contiguous or just contiguous in the FFT dim? If it's not truly row_contiguous, then we presumably need to specify the strides to the kernel?

Reply:
@barronalex FWIW, it is indeed place where nextup operations on FFT result (like `.abs()`) cause fatal error. Replacing `in_contiguous.flags()` to empty is exact that makes my experimental code working. BTW big thank you for your work on Metal FFT.

Reply:
Thanks for catching this! Will push a fix shortly.

Reply:
@barronalex would you mind adding a test or two with non-contiguous arrays (e.g. output of transpose or broadcast). The logic here is a bit subtle so some tests would be good.